lib/spack/spack/package_base.py

```diff
@@ -1418,6 +1418,18 @@ def prefix(self):
         """Get the prefix into which this package should be installed."""
         return self.spec.prefix
 
+    @property
+    def bash_completion_path(self):
+        return f"{self.prefix}/share/bash-completion/completions"
+
+    @property
+    def zsh_completion_path(self):
+        return f"{self.prefix}/share/zsh/site-functions"
+
+    @property
+    def fish_completion_path(self):
+        return f"{self.prefix}/share/fish/vendor_completions.d"
```

Comment:
Wondering if these attributes are best kept in a mixin, rather than extending the `PackageBase` class. That class is a _bit_ overloaded with responsibilities atm.

Reply:
We could totally add this in as a mixin to reduce the size of PackageBase. I'll push back a little though, since I think this fits very well in package base since it's an attribute that every package can use and it's really nice having it be globally accessible as a packager as we go and add in tab completion for more packages. (Nearly every modern cli app provides them so it'll be a mixin that's in a lot of places.)

Reply:
I don't think @alalazo was suggesting to make it a mixin that users have to enable on each package. We can encapsulate the functionality in a mixin, and then inherit from it in PackageBase. It would just encapsulate the code in one place.

I'm not sure I love the mixins -- it's a pattern we use already and it does provide good code encapsulation, but IMO it makes the code more difficult to read or spin up on.

Reply:
I see, I'd also prefer to keep it in `PackageBase`. However, I'd be alright with converting it to a mixin that is applied globally for all packages. @alalazo do you want to weigh in here as a final request to move to a mixin or approve as is? I've got a little time this week where I could patch this up if needed.

Reply:
Sorry, this one got off my radar. I was suggesting a mixing that users have to enable, but looking at it again, I think free-standing functions would be my preferred solutions. 

We can make them global (import them from `spack.package`), and avoid adding more attributes to each and every package.

> it's an attribute that every package can use [ ... ]

That's the part I disagree with. Not every package has all three, indeed I suspect that the majority of packages would use none.

Reply:
No worries, I've gone ahead and rewritten the PR to have the shell paths exposed as functions in `spack.package`.